File: session-06-capstone/go/resilience/timeout.go
package resilience

import (
	"context"
	"fmt"
	"time"

	"coffeeshop-capstone/kitchen"
)

type TimeoutKitchenService struct {
	Inner   kitchen.KitchenService
	Timeout time.Duration
}

func (s *TimeoutKitchenService) Prepare(orderID int, drink string) (*kitchen.Preparation, error) {
	ctx, cancel := context.WithTimeout(context.Background(), s.Timeout)
	defer cancel()

	type result struct {
		prep *kitchen.Preparation
		err  error
	}

	ch := make(chan result, 1)
	go func() {
		prep, err := s.Inner.Prepare(orderID, drink)
		ch <- result{prep, err}
	}()

	select {
	case r := <-ch:
		return r.prep, r.err
	case <-ctx.Done():
		return nil, fmt.Errorf("kitchen timed out after %s for order #%d", s.Timeout, orderID)
	}
}

File: session-06-capstone/python/resilience/timeout.py
import concurrent.futures

from kitchen.service import KitchenService, Preparation


class TimeoutKitchenService(KitchenService):
    def __init__(self, inner: KitchenService, timeout_seconds: float):
        self.inner = inner
        self.timeout_seconds = timeout_seconds

    def prepare(self, order_id: int, drink: str) -> Preparation:
        with concurrent.futures.ThreadPoolExecutor() as executor:
            future = executor.submit(self.inner.prepare, order_id, drink)
            try:
                return future.result(timeout=self.timeout_seconds)
            except concurrent.futures.TimeoutError:
                raise RuntimeError(
                    f"Kitchen timed out after {self.timeout_seconds}s for order #{order_id}"
                )
